src/MainWindow.vala

```diff
@@ -128,136 +125,138 @@ public class Installer.MainWindow : Gtk.Window {
 
             return Source.REMOVE;
         });
-
-        leaflet.notify["visible-child"].connect (() => {
-            update_navigation ();
-        });
-
-        leaflet.notify["child-transition-running"].connect (() => {
-            update_navigation ();
-        });
-    }
-
-    private void update_navigation () {
-        if (!leaflet.child_transition_running) {
-            // We need to rebuild the views to reflect language changes and forking paths
-            if (leaflet.visible_child == language_view || leaflet.visible_child == try_install_view) {
-                while (leaflet.get_adjacent_child (FORWARD) != null) {
-                    leaflet.remove (leaflet.get_adjacent_child (FORWARD));
-                }
-            }
-        }
     }
 
     private void load_keyboard_view () {
-        var keyboard_layout_view = new KeyboardLayoutView ();
+        keyboard_layout_view = new KeyboardLayoutView ();
         try_install_view = new TryInstallView ();
 
-        leaflet.append (keyboard_layout_view);
-        leaflet.append (try_install_view);
+        navigation_view.push (keyboard_layout_view);
 
-        leaflet.visible_child = keyboard_layout_view;
+        keyboard_layout_view.next_step.connect (() => {
+            navigation_view.push (try_install_view);
+        });
 
         try_install_view.custom_step.connect (() => {
-            load_check_view ();
-            load_partitioning_view ();
-            load_drivers_view ();
-            leaflet.navigate (FORWARD);
+            if (check_ignored) {
+                load_partitioning_view ();
+            } else {
+                load_check_view ("custom");
+            }
         });
 
         try_install_view.next_step.connect (() => {
-            load_check_view ();
-            load_disk_view ();
-            load_encrypt_view ();
-            load_drivers_view ();
-            leaflet.navigate (FORWARD);
+            if (check_ignored) {
+                load_disk_view ();
+            } else {
+                load_check_view ("standard");
+            }
         });
     }
 
     private void load_disk_view () {
         var disk_view = new DiskView ();
-        leaflet.append (disk_view);
+        navigation_view.push (disk_view);
 
         disk_view.load.begin (MINIMUM_SPACE);
-        disk_view.cancel.connect (() => leaflet.navigate (BACK));
+        disk_view.cancel.connect (() => navigation_view.pop ());
+        disk_view.next_step.connect (() => load_encrypt_view ());
     }
 
-    private void load_check_view () {
-        if (check_ignored) {
-            return;
-        }
-
+    private void load_check_view (string path) {
```

Comment:
Is there any reason you chose `string` to separate standard install and custom install? I think we may want to use a enum or bool which are more easier to handle.

Reply:
Re-wrote it without the strings!